lib/ret/application.ex

```diff
@@ -11,39 +11,54 @@ defmodule Ret.Application do
     Application.load(:ret)
     EctoBootMigration.start_dependencies()
 
-    repos_pids = EctoBootMigration.start_repos([Ret.Repo])
+    repos_pids =
+      Ret.Locking.exec_if_session_lockable("ret_migration", fn ->
+        repos_pids = EctoBootMigration.start_repos([Ret.SessionLockRepo])
 
-    Ret.Locking.exec_if_session_lockable("ret_migration", fn ->
-      db_name = Application.get_env(:ret, Ret.Repo)[:database]
+        # Note the main Repo database is used here, since the session locking database
+        # name may be a proxy database in pgbouncer which doesn't actually exist.
+        db_name = Application.get_env(:ret, Ret.Repo)[:database]
 
-      # Can't check mix_env here, so check db name
-      if db_name !== "ret_test" do
-        coturn_enabled = Ret.Coturn.enabled?()
+        # Can't check mix_env here, so check db name
+        if db_name !== "ret_test" do
+          coturn_enabled = Ret.Coturn.enabled?()
 
-        Ecto.Adapters.SQL.query!(Ret.Repo, "CREATE SCHEMA IF NOT EXISTS ret0")
+          Ecto.Adapters.SQL.query!(Ret.SessionLockRepo, "CREATE SCHEMA IF NOT EXISTS ret0")
 
-        if coturn_enabled do
-          Ecto.Adapters.SQL.query!(Ret.Repo, "CREATE SCHEMA IF NOT EXISTS coturn")
-        end
-
-        schemas =
           if coturn_enabled do
-            ["ret0", "coturn"]
-          else
-            ["ret0"]
+            Ecto.Adapters.SQL.query!(Ret.SessionLockRepo, "CREATE SCHEMA IF NOT EXISTS coturn")
+          end
+
+          Ecto.Adapters.SQL.query!(
+            Ret.SessionLockRepo,
+            "ALTER DATABASE #{db_name} SET search_path TO ret0"
+          )
+
+          priv_path = Path.join(["#{:code.priv_dir(:ret)}", "repo", "migrations"])
+
+          # Disallow stop of the application via SIGTERM until migrations are finished.
+          #
+          # If application is killed mid-migration, then it's possible for schema migrations
+          # table to not accurately reflect the migrations which have ran.
+          Ret.DelayStopSignalHandler.delay_stop()
+
+          try do
+            Ecto.Migrator.run(Ret.SessionLockRepo, priv_path, :up, all: true, prefix: "ret0")
+          after
+            Ret.DelayStopSignalHandler.allow_stop()
           end
```

Comment:
I believe we won't return `repos_pids` if there's an error in the `try`. Should we also `stop_repos` in the `after` clause?

Reply:
`after` is like `finally`, if it throws in try this will throw, and hence stop starting the application altogether. (since this is the application start process)